hotshot-state-prover/src/service.rs

```diff
@@ -315,7 +340,39 @@ pub async fn submit_state_and_proof(
     // prepare the input the contract call and the tx itself
     let proof: ParsedPlonkProof = proof.into();
     let new_state: ParsedLightClientState = public_input.into();
-    let tx = contract.new_finalized_state(new_state.into(), proof.into());
+
+    let mut tx = contract.new_finalized_state(new_state.into(), proof.into());
+
+    // TODO: (alex) centralize this to a dedicated ethtxmanager with gas control
+    // frugal gas price: set to SafeGasPrice based on live price oracle
+    // safe to be included with low priority position in a block.
+    // ignore this if etherscan fail to return
+    if let Ok(res) =
+        reqwest::get("https://api.etherscan.io/api?module=gastracker&action=gasoracle").await
+    {
+        // parse etherscan response,
+        // if etherscan API changes, we log the warning and fallback to default (by falling through)
+        if let Ok(res) = res.json::<EtherscanPriceOracleResponse>().await {
+            if let TypedTransaction::Eip1559(inner) = &mut tx.tx {
+                // both safe unwraps below assuming correct etherscan stream
+                let safe_max_fee: U256 = parse_units(res.result.safe_max_fee.clone(), "gwei")
+                    .unwrap()
+                    .into();
+                let base_fee: U256 = parse_units(res.result.base_fee.clone(), "gwei")
+                    .unwrap()
```

Comment:
Suggest removing unwraps, maybe add a helper function to parse the API response into a type that is convenient for us to work with.

Reply:
should we just return error (propagate err up) or still use unwrap but hide inside some helper function?